File: backend/app/services/note_service.py
from sqlalchemy.orm import Session

from math import ceil
from datetime import datetime, timezone

from app.models import Note
from app.schemas.note import NoteCreate, NoteUpdate
from app.repositories import note_repository


def create_note(
    db: Session,
    note_data: NoteCreate,
    user_id: int,
) -> Note:

    note = Note(
        user_id=user_id,
        title=note_data.title,
        content=note_data.content,
    )

    if note_data.tag_ids:
        tags = note_repository.get_tags_by_ids(
            db,
            note_data.tag_ids,
            user_id,
        )

        if len(tags) != len(set(note_data.tag_ids)):
            raise ValueError(
                "One or more tag IDs do not exist"
            )

        note.tags = tags

    return note_repository.create_note(
        db,
        note,
    )


def get_notes(
    db: Session,
    user_id: int,
    favorite: bool | None = None,
    deleted: bool = False,
    tag_id: int | None = None,
    query: str | None = None,
    page: int = 1,
    limit: int = 12,
):

    notes, total = note_repository.get_notes(
        db=db,
        user_id=user_id,
        favorite=favorite,
        deleted=deleted,
        tag_id=tag_id,
        query=query,
        page=page,
        limit=limit,
    )

    total_pages = ceil(total / limit) if total > 0 else 0

    return {
        "items": notes,
        "total": total,
        "page": page,
        "limit": limit,
        "total_pages": total_pages,
    }


def get_note(
    db: Session,
    note_id: int,
    user_id: int,
) -> Note | None:

    return note_repository.get_note(
        db,
        note_id,
        user_id,
    )


def update_note(
    db: Session,
    note_id: int,
    user_id: int,
    note_data: NoteUpdate,
) -> Note | None:

    note = note_repository.get_note(
        db,
        note_id,
        user_id,
    )

    if note is None or note.is_deleted:
        return None

    content_updated = False

    if note_data.title is not None:
        note.title = note_data.title
        content_updated = True

    if note_data.content is not None:
        note.content = note_data.content
        content_updated = True

    if note_data.tag_ids is not None:

        if note_data.tag_ids:
            tags = note_repository.get_tags_by_ids(
                db,
                note_data.tag_ids,
                user_id,
            )

            if len(tags) != len(
                set(note_data.tag_ids)
            ):
                raise ValueError(
                    "One or more tag IDs do not exist"
                )

            note.tags = tags

        else:
            note.tags = []

        content_updated = True

    if note_data.is_favorite is not None:
        note.is_favorite = note_data.is_favorite

    if content_updated:
        note.updated_at = datetime.now(timezone.utc)

    return note_repository.update_note(
        db,
        note,
    )


def soft_delete_note(
    db: Session,
    note_id: int,
    user_id: int,
) -> Note | None:

    note = note_repository.get_note(
        db,
        note_id,
        user_id,
    )

    if note is None or note.is_deleted:
        return None

    note.is_deleted = True

    return note_repository.update_note(
        db,
        note,
    )


def restore_note(
    db: Session,
    note_id: int,
    user_id: int,
) -> Note | None:

    note = note_repository.get_note(
        db,
        note_id,
        user_id,
    )

    if note is None or not note.is_deleted:
        return None

    note.is_deleted = False

    return note_repository.update_note(
        db,
        note,
    )


def hard_delete_note(
    db: Session,
    note_id: int,
    user_id: int,
) -> bool:

    note = note_repository.get_note(
        db,
        note_id,
        user_id,
    )

    if note is None or not note.is_deleted:
        return False

    note_repository.hard_delete_note(
        db,
        note,
    )

    return True

File: frontend/src/services/noteService.ts
import api from "./api"

import type {
  TagWithCount,
} from "./tagService"

export type Note = {
  id: number
  user_id: number
  title: string
  content: string

  is_favorite: boolean
  is_deleted: boolean

  tags: TagWithCount[]

  created_at: string
  updated_at: string
}

export type PaginatedNotesResponse = {
  items: Note[]
  total: number
  page: number
  limit: number
  total_pages: number
}

type GetNotesParams = {
  favorite?: boolean
  deleted?: boolean
  tag_id?: number
  q?: string
  page?: number
  limit?: number
}

type CreateNoteData = {
  title: string
  content: string
  tag_ids?: number[]
}

type UpdateNoteData = {
  title?: string
  content?: string
  tag_ids?: number[]
  is_favorite?: boolean
}

export const getNotes = async (
  params: GetNotesParams = {},
): Promise<PaginatedNotesResponse> => {
  const searchParams =
    new URLSearchParams()

  if (params.favorite !== undefined) {
    searchParams.set(
      "favorite",
      String(params.favorite),
    )
  }

  if (params.deleted !== undefined) {
    searchParams.set(
      "deleted",
      String(params.deleted),
    )
  }

  if (params.tag_id !== undefined) {
    searchParams.set(
      "tag_id",
      String(params.tag_id),
    )
  }

  if (params.q) {
    searchParams.set(
      "q",
      params.q,
    )
  }

  if (params.page !== undefined) {
    searchParams.set(
      "page",
      String(params.page),
    )
  }

  if (params.limit !== undefined) {
    searchParams.set(
      "limit",
      String(params.limit),
    )
  }

  const query =
    searchParams.toString()

  return api<PaginatedNotesResponse>(
    `/api/notes${
      query ? `?${query}` : ""
    }`,
  )
}

export const getNote = async (
  noteId: number,
): Promise<Note> => {
  return api<Note>(
    `/api/notes/${noteId}`,
  )
}

export const createNote = async (
  data: CreateNoteData,
): Promise<Note> => {
  return api<Note>(
    "/api/notes",
    {
      method: "POST",
      body: JSON.stringify(data),
    },
  )
}

export const updateNote = async (
  noteId: number,
  data: UpdateNoteData,
): Promise<Note> => {
  return api<Note>(
    `/api/notes/${noteId}`,
    {
      method: "PATCH",
      body: JSON.stringify(data),
    },
  )
}

export const deleteNote = async (
  noteId: number,
): Promise<Note> => {
  return api<Note>(
    `/api/notes/${noteId}`,
    {
      method: "DELETE",
    },
  )
}

export const restoreNote = async (
  noteId: number,
): Promise<Note> => {
  return api<Note>(
    `/api/notes/${noteId}/restore`,
    {
      method: "PATCH",
    },
  )
}

export const permanentlyDeleteNote = async (
  noteId: number,
): Promise<void> => {
  return api<void>(
    `/api/notes/${noteId}/permanent`,
    {
      method: "DELETE",
    },
  )
}